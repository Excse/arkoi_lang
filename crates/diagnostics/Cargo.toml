[package]
name = "diagnostics"
version = "0.1.0"
edition = "2021"

[dependencies]
colored = "2.0.4"
serde = { version = "1.0.188", features = ["derive"] }
strum = "0.25"
strum_macros = "0.25"
