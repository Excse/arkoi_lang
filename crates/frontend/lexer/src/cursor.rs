#[cfg(feature = "serialize")]
use serde::Serialize;

use std::{iter::Peekable, str::CharIndices};

use crate::error::{DidntExpect, EndOfFile, Result};
use diagnostics::{
    file::{FileID, Files},
    positional::{LabelSpan, Span},
};

#[cfg_attr(feature = "serialize", derive(Serialize))]
#[derive(Debug)]
pub struct Cursor<'a> {
    file_id: FileID,
    files: &'a Files,
    #[serde(skip)]
    chars: Peekable<CharIndices<'a>>,
    length: usize,
    start: usize,
}

impl<'a> Cursor<'a> {
    pub fn new(file_id: FileID, files: &'a Files) -> Cursor<'a> {
        let source = files
            .source(file_id)
            .expect("Couldn't get the source of this file.");

        Self {
            file_id,
            files,
            chars: source.char_indices().peekable(),
            length: source.len(),
            start: 0,
        }
    }

    pub fn current_index(&mut self) -> usize {
        self.peek_indexed()
            .map(|(index, _)| index)
            .unwrap_or(self.length)
    }

    pub fn mark_start(&mut self) {
        self.start = self.current_index()
    }

    pub fn as_span(&mut self) -> LabelSpan {
        LabelSpan::new(Span::new(self.start, self.current_index()), self.file_id)
    }

    // TODO: Remove the expect
    pub fn as_str(&mut self) -> &'a str {
        let span = self.as_span();
        self.files
            .slice(self.file_id, &span.span)
            .expect("Couldn't slice the source")
    }

    pub fn peek_indexed(&mut self) -> Option<(usize, char)> {
        self.chars.peek().copied()
    }

    pub fn peek(&mut self) -> Option<char> {
        self.peek_indexed().map(|(_, char)| char)
    }

    pub fn try_consume(&mut self) -> Option<char> {
        let char = self.chars.next().map(|(_, char)| char)?;
        Some(char)
    }

    pub fn try_eat(&mut self, expected: char) -> Result<char> {
        match self.peek_indexed() {
            Some((_, char)) if char == expected => Ok(self.try_consume().unwrap()),
            Some((index, char)) => Err(DidntExpect::new(
                char,
                LabelSpan::new(Span::single(index), self.file_id),
                expected,
            )
            .into()),
            None => Err(EndOfFile.into()),
        }
    }

    pub fn eat_if<F>(&mut self, predicate: F, message: &'static str) -> Result<char>
    where
        F: FnOnce(char) -> bool,
    {
        match self.peek_indexed() {
            Some((_, char)) if predicate(char) => Ok(self.try_consume().unwrap()),
            Some((index, char)) => Err(DidntExpect::new(
                char,
                LabelSpan::new(Span::single(index), self.file_id),
                message,
            )
            .into()),
            None => Err(EndOfFile.into()),
        }
    }

    pub fn eat_while<F>(&mut self, mut predicate: F)
    where
        F: FnMut(char) -> bool,
    {
        self.eat_windowed_while(|_, current| predicate(current))
    }

    pub fn eat_windowed_while<F>(&mut self, mut predicate: F)
    where
        F: FnMut(char, char) -> bool,
    {
        let mut last: char = '\0';
        while let Some(char) = self.peek() {
            if !predicate(last, char) {
                break;
            }

            last = char;
            self.try_consume();
        }
    }
}
