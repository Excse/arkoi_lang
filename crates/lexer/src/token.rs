use serde::Serialize;
use serdebug::SerDebug;
use strum::AsRefStr;

use diagnostics::Span;

#[derive(SerDebug, Serialize)]
pub struct Token<'a> {
    pub span: Span<'a>,
    pub value: Option<TokenValue<'a>>,
    pub kind: TokenKind,
}

#[derive(SerDebug, Serialize)]
pub enum TokenValue<'a> {
    Integer(usize),
    Decimal(f64),
    String(&'a str),
    Boolean(bool),
}

#[derive(SerDebug, Serialize, Eq, PartialEq, Copy, Clone, AsRefStr)]
pub enum TokenKind {
    #[serde(rename = "int")]
    #[strum(serialize = "int")]
    Integer,
    #[serde(rename = "decimal")]
    #[strum(serialize = "decimal")]
    Decimal,
    #[serde(rename = "identifier")]
    #[strum(serialize = "identifier")]
    Identifier,
    #[serde(rename = "string")]
    #[strum(serialize = "string")]
    String,
    #[serde(rename = "true")]
    #[strum(serialize = "true")]
    True,
    #[serde(rename = "false")]
    #[strum(serialize = "false")]
    False,

    #[serde(rename = "struct")]
    #[strum(serialize = "struct")]
    Struct,
    #[serde(rename = "fun")]
    #[strum(serialize = "fun")]
    Fun,
    #[serde(rename = "let")]
    #[strum(serialize = "let")]
    Let,
    #[serde(rename = "return")]
    #[strum(serialize = "return")]
    Return,

    #[serde(rename = "{")]
    #[strum(serialize = "{")]
    OBracket,
    #[serde(rename = "}")]
    #[strum(serialize = "}")]
    CBracket,
    #[serde(rename = "(")]
    #[strum(serialize = "(")]
    OParent,
    #[serde(rename = ")")]
    #[strum(serialize = ")")]
    CParent,
    #[serde(rename = "@")]
    #[strum(serialize = "@")]
    At,
    #[serde(rename = "!")]
    #[strum(serialize = "|")]
    Apostrophe,
    #[serde(rename = ",")]
    #[strum(serialize = ",")]
    Comma,
    #[serde(rename = ".")]
    #[strum(serialize = ".")]
    Period,
    #[serde(rename = ";")]
    #[strum(serialize = ";")]
    Semicolon,

    #[serde(rename = "+=")]
    #[strum(serialize = "+=")]
    AddAssign,
    #[serde(rename = "+")]
    #[strum(serialize = "+=")]
    Plus,
    #[serde(rename = "-=")]
    #[strum(serialize = "-=")]
    MinusAssign,
    #[serde(rename = "-")]
    #[strum(serialize = "-")]
    Minus,
    #[serde(rename = "*=")]
    #[strum(serialize = "*=")]
    MulAssign,
    #[serde(rename = "*")]
    #[strum(serialize = "*")]
    Asterisk,
    #[serde(rename = "/=")]
    #[strum(serialize = "/=")]
    DivAssign,
    #[serde(rename = "/")]
    #[strum(serialize = "/")]
    Slash,
    #[serde(rename = "<=")]
    #[strum(serialize = "<=")]
    LessEqual,
    #[serde(rename = "<")]
    #[strum(serialize = "<")]
    Less,
    #[serde(rename = ">=")]
    #[strum(serialize = ">=")]
    GreaterEqual,
    #[serde(rename = ">")]
    #[strum(serialize = ">")]
    Greater,
    #[serde(rename = "==")]
    #[strum(serialize = "==")]
    Equal,
    #[serde(rename = "!=")]
    #[strum(serialize = "!=")]
    NotEqual,
    #[serde(rename = "=")]
    #[strum(serialize = "=")]
    Assign,

    #[serde(rename = "self")]
    #[strum(serialize = "self")]
    Self_,
    #[serde(rename = "u8")]
    #[strum(serialize = "u8")]
    U8,
    #[serde(rename = "i8")]
    #[strum(serialize = "i8")]
    I8,
    #[serde(rename = "u16")]
    #[strum(serialize = "u16")]
    U16,
    #[serde(rename = "i16")]
    #[strum(serialize = "i16")]
    I16,
    #[serde(rename = "u32")]
    #[strum(serialize = "u32")]
    U32,
    #[serde(rename = "i32")]
    #[strum(serialize = "i32")]
    I32,
    #[serde(rename = "u64")]
    #[strum(serialize = "u64")]
    U64,
    #[serde(rename = "i64")]
    #[strum(serialize = "i64")]
    I64,
    #[serde(rename = "usize")]
    #[strum(serialize = "usize")]
    USize,
    #[serde(rename = "isize")]
    #[strum(serialize = "isize")]
    ISize,
    #[serde(rename = "f32")]
    #[strum(serialize = "f32")]
    F32,
    #[serde(rename = "f64")]
    #[strum(serialize = "f64")]
    F64,

    #[serde(rename = "unknown")]
    #[strum(serialize = "unknown")]
    Unknown,
}

impl<'a> Token<'a> {
    pub fn new(span: Span<'a>, value: Option<TokenValue<'a>>, kind: TokenKind) -> Token<'a> {
        Token { span, value, kind }
    }
}
