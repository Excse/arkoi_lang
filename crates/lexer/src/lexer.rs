#[cfg(feature = "serialize")]
use serde::Serialize;

use lasso::Rodeo;

use crate::cursor::Cursor;
use crate::token::{Token, TokenKind, TokenValue};
use diagnostics::{
    file::{FileID, Files},
    report::Report,
};

#[cfg_attr(feature = "serialize", derive(Serialize))]
#[derive(Debug)]
pub struct Lexer<'a> {
    cursor: Cursor<'a>,
    interner: &'a mut Rodeo,
    pub errors: Vec<LexerError>,
}

#[cfg_attr(feature = "serialize", derive(Serialize))]
#[derive(Debug)]
pub enum LexerError {
    Diagnostic(Report),
    EndOfFile,
}

#[cfg_attr(feature = "serialize", derive(Serialize))]
#[derive(Debug)]
pub struct TokenIter<'a> {
    lexer: &'a mut Lexer<'a>,
}

impl<'a> Iterator for TokenIter<'a> {
    type Item = Token;

    fn next(&mut self) -> Option<Self::Item> {
        let token_kind = match self.lexer.next_token_kind() {
            Ok(token_kind) => token_kind,
            Err(error) => match error {
                LexerError::Diagnostic(_) => {
                    self.lexer.errors.push(error);
                    return self.next();
                }
                _ => return None,
            },
        };

        let content = self.lexer.cursor.as_str();
        let span = self.lexer.cursor.as_span();

        let value = match token_kind {
            TokenKind::Integer => Some(TokenValue::Integer(content.parse::<usize>().unwrap())),
            TokenKind::Decimal => Some(TokenValue::Decimal(content.parse::<f64>().unwrap())),
            TokenKind::Identifier => Some(TokenValue::String(
                self.lexer.interner.get_or_intern(content),
            )),
            TokenKind::String => Some(TokenValue::String(
                self.lexer
                    .interner
                    .get_or_intern(&content[1..content.len() - 1]),
            )),
            TokenKind::True => Some(TokenValue::Boolean(true)),
            TokenKind::False => Some(TokenValue::Boolean(false)),
            _ => None,
        };

        Some(Token::new(span, value, token_kind))
    }
}

impl<'a> Lexer<'a> {
    pub fn new(files: &'a Files, file_id: FileID, interner: &'a mut Rodeo) -> Lexer<'a> {
        Lexer {
            cursor: Cursor::new(file_id, files),
            interner,
            errors: Vec::new(),
        }
    }

    pub fn iter(&'a mut self) -> TokenIter<'a> {
        TokenIter { lexer: self }
    }

    fn next_token_kind(&mut self) -> Result<TokenKind, LexerError> {
        self.cursor.mark_start();
        match self.cursor.peek() {
            Some(char) if char.is_alphabetic() => self.read_identifier(),
            Some(char) if char.is_numeric() => self.read_number(),
            Some('"') => self.read_string(),
            Some(_) => self.read_symbol(),
            None => Err(LexerError::EndOfFile),
        }
    }

    fn read_symbol(&mut self) -> Result<TokenKind, LexerError> {
        let mut token = match self.cursor.consume() {
            Some(char) if char.is_whitespace() => self.next_token_kind()?,
            Some('{') => TokenKind::OBracket,
            Some('}') => TokenKind::CBracket,
            Some('(') => TokenKind::OParent,
            Some(')') => TokenKind::CParent,
            Some('@') => TokenKind::At,
            Some(',') => TokenKind::Comma,
            Some('.') => TokenKind::Period,
            Some('+') => TokenKind::Plus,
            Some('-') => TokenKind::Minus,
            Some('*') => TokenKind::Asterisk,
            Some('/') => TokenKind::Slash,
            Some('<') => TokenKind::Less,
            Some('>') => TokenKind::Greater,
            Some('=') => TokenKind::Assign,
            Some('!') => TokenKind::Apostrophe,
            Some(';') => TokenKind::Semicolon,
            _ => TokenKind::Unknown,
        };

        let current = match self.cursor.peek() {
            Some(char) => char,
            None => return Ok(token),
        };

        token = match (token, current) {
            (TokenKind::Plus, '=') => TokenKind::AddAssign,
            (TokenKind::Minus, '=') => TokenKind::MinusAssign,
            (TokenKind::Asterisk, '=') => TokenKind::MulAssign,
            (TokenKind::Slash, '=') => TokenKind::DivAssign,
            (TokenKind::Less, '=') => TokenKind::LessEqual,
            (TokenKind::Greater, '=') => TokenKind::GreaterEqual,
            (TokenKind::Assign, '=') => TokenKind::Equal,
            (TokenKind::Apostrophe, '=') => TokenKind::NotEqual,
            (token, _) => return Ok(token),
        };

        self.cursor.consume();

        Ok(token)
    }

    fn read_identifier(&mut self) -> Result<TokenKind, LexerError> {
        self.cursor.eat_if(char::is_alphabetic, "a-zA-Z")?;

        self.cursor.eat_while(char::is_alphanumeric);

        Ok(match self.cursor.as_str() {
            "true" => TokenKind::True,
            "false" => TokenKind::False,
            "struct" => TokenKind::Struct,
            "return" => TokenKind::Return,
            "let" => TokenKind::Let,
            "self" => TokenKind::Self_,
            "fun" => TokenKind::Fun,
            "u8" => TokenKind::U8,
            "i8" => TokenKind::I8,
            "u16" => TokenKind::U16,
            "i16" => TokenKind::I16,
            "u32" => TokenKind::U32,
            "i32" => TokenKind::I32,
            "u64" => TokenKind::U64,
            "i64" => TokenKind::I64,
            "usize" => TokenKind::USize,
            "isize" => TokenKind::ISize,
            "f32" => TokenKind::F32,
            "f64" => TokenKind::F64,
            _ => TokenKind::Identifier,
        })
    }

    fn read_number(&mut self) -> Result<TokenKind, LexerError> {
        self.cursor.eat_if(char::is_numeric, "0-9")?;

        self.cursor.eat_while(char::is_numeric);

        if self.cursor.eat('.').is_ok() {
            self.cursor.eat_while(char::is_numeric);
            Ok(TokenKind::Decimal)
        } else {
            Ok(TokenKind::Integer)
        }
    }

    fn read_string(&mut self) -> Result<TokenKind, LexerError> {
        self.cursor.eat('"')?;

        self.cursor
            .eat_windowed_while(|prev, curr| curr != '"' || prev == '\\');

        self.cursor.eat('"')?;

        Ok(TokenKind::String)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    macro_rules! test_token {
        (FAIL: $name:ident, $func:ident, $source:expr) => {
            #[test]
            fn $name() {
                let mut files = Files::default();
                let file_id = files.add("test.ark", $source);

                let mut interner = Rodeo::default();

                let mut lexer = Lexer::new(&files, file_id, &mut interner);
                let token = lexer.$func();
                assert!(token.is_err(), "{:?} should be an error", token);
            }
        };
        ($name:ident, $source:expr => $expected:expr) => {
            #[test]
            fn $name() {
                let mut files = Files::default();
                let file_id = files.add("test.ark", $source);

                let mut interner = Rodeo::default();

                let mut lexer = Lexer::new(&files, file_id, &mut interner);
                let expected = TokenKind::from($expected);
                let token = lexer.next_token_kind().unwrap();
                assert!(token == expected, "Input was {:?}", $source);
            }
        };
    }

    test_token!(success_decimal, "4.2" => TokenKind::Decimal);
    test_token!(success_integer, "42" => TokenKind::Integer);
    test_token!(FAIL: fail_number, read_number, "number");

    test_token!(success_string, "\"Hello World!\"" => TokenKind::String);
    test_token!(FAIL: fail_string, read_string, "Hello World!");

    test_token!(success_true, "true" => TokenKind::True);
    test_token!(success_false, "false" => TokenKind::False);

    test_token!(success_obracket, "{" => TokenKind::OBracket);
    test_token!(success_cbracket, "}" => TokenKind::CBracket);
    test_token!(success_oparent, "(" => TokenKind::OParent);
    test_token!(success_cparent, ")" => TokenKind::CParent);
    test_token!(success_at, "@" => TokenKind::At);
    test_token!(success_apostrophe, "!" => TokenKind::Apostrophe);
    test_token!(success_comma, "," => TokenKind::Comma);
    test_token!(success_period, "." => TokenKind::Period);
    test_token!(success_semicolon, ";" => TokenKind::Semicolon);
    test_token!(success_addassign, "+=" => TokenKind::AddAssign);
    test_token!(success_plus, "+" => TokenKind::Plus);
    test_token!(success_minusassing, "-=" => TokenKind::MinusAssign);
    test_token!(success_minus, "-" => TokenKind::Minus);
    test_token!(success_mulassign, "*=" => TokenKind::MulAssign);
    test_token!(success_asterisk, "*" => TokenKind::Asterisk);
    test_token!(success_divassign, "/=" => TokenKind::DivAssign);
    test_token!(success_slash, "/" => TokenKind::Slash);
    test_token!(success_lessequal, "<=" => TokenKind::LessEqual);
    test_token!(success_less, "<" => TokenKind::Less);
    test_token!(success_greaterequal, ">=" => TokenKind::GreaterEqual);
    test_token!(success_greater, ">" => TokenKind::Greater);
    test_token!(success_equal, "==" => TokenKind::Equal);
    test_token!(success_notequal, "!=" => TokenKind::NotEqual);
    test_token!(success_assign, "=" => TokenKind::Assign);

    test_token!(success_self, "self" => TokenKind::Self_);
    test_token!(success_u8, "u8" => TokenKind::U8);
    test_token!(success_i8, "i8" => TokenKind::I8);
    test_token!(success_u16, "u16" => TokenKind::U16);
    test_token!(success_i16, "i16" => TokenKind::I16);
    test_token!(success_u32, "u32" => TokenKind::U32);
    test_token!(success_i32, "i32" => TokenKind::I32);
    test_token!(success_u64, "u64" => TokenKind::U64);
    test_token!(success_i64, "i64" => TokenKind::I64);
    test_token!(success_usize, "usize" => TokenKind::USize);
    test_token!(success_isize, "isize" => TokenKind::ISize);
    test_token!(success_f32, "f32" => TokenKind::F32);
    test_token!(success_f64, "f64" => TokenKind::F64);

    macro_rules! insta_test {
        ($name:ident, $path:expr) => {
            #[derive(Serialize)]
            struct InstaSnapshot<'a> {
                tokens: &'a [Token],
                interner: &'a Rodeo,
            }

            #[test]
            fn $name() {
                let mut files = Files::default();

                let source = std::fs::read_to_string($path).expect("Couldn't read the file.");
                let file_id = files.add($path, &source);

                let mut interner = Rodeo::default();

                let mut lexer = Lexer::new(&files, file_id, &mut interner);
                let tokens = lexer.iter().collect::<Vec<Token>>();

                insta::assert_yaml_snapshot!(InstaSnapshot {
                    tokens: &tokens,
                    interner: &interner,
                });
            }
        };
    }

    insta_test!(insta_test, "test_files/insta_test.ark");
}
