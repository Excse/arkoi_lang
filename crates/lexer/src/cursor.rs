#[cfg(feature = "serialize")]
use serde::Serialize;

use std::str::Chars;
use std::{iter::Peekable, str::CharIndices};

use crate::lexer::LexerError;
use diagnostics::{
    file::{FileID, Files},
    positional::{Span, Spannable},
};
use errors::lexer::*;

#[cfg_attr(feature = "serialize", derive(Serialize))]
#[derive(Debug)]
pub struct Cursor<'a> {
    file_id: FileID,
    files: &'a Files,
    #[serde(skip)]
    chars: Peekable<CharIndices<'a>>,
    length: usize,
    start: usize,
}

impl<'a> Cursor<'a> {
    pub fn new(file_id: FileID, files: &'a Files) -> Cursor<'a> {
        let source = files
            .source(file_id)
            .expect("Couldn't get the source of this file.");

        Cursor {
            file_id,
            files,
            chars: source.char_indices().peekable(),
            length: source.len(),
            start: 0,
        }
    }

    pub fn current_index(&mut self) -> usize {
        self.peek_indexed()
            .map(|(index, _)| index)
            .unwrap_or(self.length)
    }

    pub fn mark_start(&mut self) {
        self.start = self.current_index()
    }

    pub fn as_span(&mut self) -> Span {
        Span::new(self.start, self.current_index())
    }

    // TODO: Remove the expect
    pub fn as_str(&mut self) -> &'a str {
        let span = self.as_span();
        self.files
            .slice(self.file_id, &span)
            .expect("Couldn't slice the source")
    }

    pub fn peek_indexed(&mut self) -> Option<(usize, char)> {
        self.chars.peek().copied()
    }

    pub fn peek(&mut self) -> Option<char> {
        self.peek_indexed().map(|(_, char)| char)
    }

    pub fn try_consume(&mut self) -> Option<char> {
        let char = self.chars.next().map(|(_, char)| char)?;
        Some(char)
    }

    pub fn try_eat(&mut self, expected: char) -> Result<char, LexerError> {
        match self.peek_indexed() {
            Some((_, char)) if char == expected => Ok(self.try_consume().unwrap()),
            Some((index, char)) => Err(LexerError::Diagnostic(didnt_expect(
                self.files,
                self.file_id,
                Spannable::new(char, Span::empty(index)),
                expected.to_string(),
            ))),
            None => Err(LexerError::EndOfFile),
        }
    }

    pub fn eat_if<F>(&mut self, predicate: F, message: &'static str) -> Result<char, LexerError>
    where
        F: FnOnce(char) -> bool,
    {
        match self.peek_indexed() {
            Some((_, char)) if predicate(char) => Ok(self.try_consume().unwrap()),
            Some((index, char)) => Err(LexerError::Diagnostic(didnt_expect(
                self.files,
                self.file_id,
                Spannable::new(char, Span::empty(index)),
                message.to_string(),
            ))),
            None => Err(LexerError::EndOfFile),
        }
    }

    pub fn eat_while<F>(&mut self, mut predicate: F)
    where
        F: FnMut(char) -> bool,
    {
        self.eat_windowed_while(|_, current| predicate(current))
    }

    pub fn eat_windowed_while<F>(&mut self, mut predicate: F)
    where
        F: FnMut(char, char) -> bool,
    {
        let mut last: char = '\0';
        while let Some(char) = self.peek() {
            if !predicate(last, char) {
                break;
            }

            last = char;
            self.try_consume();
        }
    }
}
