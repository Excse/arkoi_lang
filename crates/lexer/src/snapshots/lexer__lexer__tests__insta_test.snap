---
source: crates/lexer/src/lexer.rs
expression: "InstaSnapshot { tokens: &tokens, interner: &interner }"
---
tokens:
  - span:
      start: 0
      end: 6
    file_id: 1
    value: ~
    kind: struct
  - span:
      start: 7
      end: 10
    file_id: 1
    value:
      String: 1
    kind: identifier
  - span:
      start: 11
      end: 12
    file_id: 1
    value: ~
    kind: "{"
  - span:
      start: 14
      end: 17
    file_id: 1
    value:
      String: 2
    kind: identifier
  - span:
      start: 18
      end: 19
    file_id: 1
    value: ~
    kind: "@"
  - span:
      start: 19
      end: 22
    file_id: 1
    value: ~
    kind: i32
  - span:
      start: 22
      end: 23
    file_id: 1
    value: ~
    kind: "!"
  - span:
      start: 23
      end: 24
    file_id: 1
    value: ~
    kind: ","
  - span:
      start: 26
      end: 29
    file_id: 1
    value:
      String: 3
    kind: identifier
  - span:
      start: 30
      end: 31
    file_id: 1
    value: ~
    kind: "@"
  - span:
      start: 31
      end: 34
    file_id: 1
    value: ~
    kind: i64
  - span:
      start: 34
      end: 35
    file_id: 1
    value: ~
    kind: ","
  - span:
      start: 36
      end: 37
    file_id: 1
    value: ~
    kind: "}"
  - span:
      start: 39
      end: 42
    file_id: 1
    value: ~
    kind: fun
  - span:
      start: 43
      end: 47
    file_id: 1
    value:
      String: 4
    kind: identifier
  - span:
      start: 47
      end: 48
    file_id: 1
    value: ~
    kind: (
  - span:
      start: 48
      end: 52
    file_id: 1
    value: ~
    kind: self
  - span:
      start: 53
      end: 54
    file_id: 1
    value: ~
    kind: "@"
  - span:
      start: 54
      end: 57
    file_id: 1
    value:
      String: 1
    kind: identifier
  - span:
      start: 57
      end: 58
    file_id: 1
    value: ~
    kind: "!"
  - span:
      start: 58
      end: 59
    file_id: 1
    value: ~
    kind: )
  - span:
      start: 60
      end: 61
    file_id: 1
    value: ~
    kind: "@"
  - span:
      start: 61
      end: 65
    file_id: 1
    value:
      String: 5
    kind: identifier
  - span:
      start: 66
      end: 67
    file_id: 1
    value: ~
    kind: "{"
  - span:
      start: 69
      end: 73
    file_id: 1
    value: ~
    kind: self
  - span:
      start: 73
      end: 74
    file_id: 1
    value: ~
    kind: "."
  - span:
      start: 74
      end: 77
    file_id: 1
    value:
      String: 2
    kind: identifier
  - span:
      start: 78
      end: 79
    file_id: 1
    value: ~
    kind: "="
  - span:
      start: 80
      end: 82
    file_id: 1
    value:
      Integer: 42
    kind: int
  - span:
      start: 82
      end: 83
    file_id: 1
    value: ~
    kind: ;
  - span:
      start: 84
      end: 85
    file_id: 1
    value: ~
    kind: "}"
  - span:
      start: 87
      end: 93
    file_id: 1
    value: ~
    kind: struct
  - span:
      start: 94
      end: 98
    file_id: 1
    value:
      String: 6
    kind: identifier
  - span:
      start: 99
      end: 100
    file_id: 1
    value: ~
    kind: "{"
  - span:
      start: 103
      end: 108
    file_id: 1
    value:
      String: 7
    kind: identifier
  - span:
      start: 109
      end: 110
    file_id: 1
    value: ~
    kind: "@"
  - span:
      start: 110
      end: 113
    file_id: 1
    value: ~
    kind: i64
  - span:
      start: 113
      end: 114
    file_id: 1
    value: ~
    kind: "!"
  - span:
      start: 114
      end: 115
    file_id: 1
    value: ~
    kind: ","
  - span:
      start: 117
      end: 122
    file_id: 1
    value:
      String: 8
    kind: identifier
  - span:
      start: 123
      end: 124
    file_id: 1
    value: ~
    kind: "@"
  - span:
      start: 124
      end: 127
    file_id: 1
    value: ~
    kind: i64
  - span:
      start: 127
      end: 128
    file_id: 1
    value: ~
    kind: "!"
  - span:
      start: 128
      end: 129
    file_id: 1
    value: ~
    kind: ","
  - span:
      start: 131
      end: 134
    file_id: 1
    value:
      String: 2
    kind: identifier
  - span:
      start: 135
      end: 136
    file_id: 1
    value: ~
    kind: "@"
  - span:
      start: 136
      end: 139
    file_id: 1
    value:
      String: 1
    kind: identifier
  - span:
      start: 139
      end: 140
    file_id: 1
    value: ~
    kind: ","
  - span:
      start: 141
      end: 142
    file_id: 1
    value: ~
    kind: "}"
  - span:
      start: 144
      end: 147
    file_id: 1
    value: ~
    kind: fun
  - span:
      start: 148
      end: 152
    file_id: 1
    value:
      String: 4
    kind: identifier
  - span:
      start: 152
      end: 153
    file_id: 1
    value: ~
    kind: (
  - span:
      start: 153
      end: 157
    file_id: 1
    value: ~
    kind: self
  - span:
      start: 158
      end: 159
    file_id: 1
    value: ~
    kind: "@"
  - span:
      start: 159
      end: 163
    file_id: 1
    value:
      String: 6
    kind: identifier
  - span:
      start: 163
      end: 164
    file_id: 1
    value: ~
    kind: "!"
  - span:
      start: 164
      end: 165
    file_id: 1
    value: ~
    kind: )
  - span:
      start: 166
      end: 167
    file_id: 1
    value: ~
    kind: "@"
  - span:
      start: 167
      end: 171
    file_id: 1
    value:
      String: 5
    kind: identifier
  - span:
      start: 172
      end: 173
    file_id: 1
    value: ~
    kind: "{"
  - span:
      start: 175
      end: 179
    file_id: 1
    value: ~
    kind: self
  - span:
      start: 179
      end: 180
    file_id: 1
    value: ~
    kind: "."
  - span:
      start: 180
      end: 185
    file_id: 1
    value:
      String: 7
    kind: identifier
  - span:
      start: 186
      end: 187
    file_id: 1
    value: ~
    kind: "="
  - span:
      start: 188
      end: 190
    file_id: 1
    value:
      Integer: 21
    kind: int
  - span:
      start: 190
      end: 191
    file_id: 1
    value: ~
    kind: ;
  - span:
      start: 193
      end: 197
    file_id: 1
    value: ~
    kind: self
  - span:
      start: 197
      end: 198
    file_id: 1
    value: ~
    kind: "."
  - span:
      start: 198
      end: 203
    file_id: 1
    value:
      String: 8
    kind: identifier
  - span:
      start: 204
      end: 205
    file_id: 1
    value: ~
    kind: "="
  - span:
      start: 206
      end: 208
    file_id: 1
    value:
      Integer: 21
    kind: int
  - span:
      start: 208
      end: 209
    file_id: 1
    value: ~
    kind: ;
  - span:
      start: 210
      end: 211
    file_id: 1
    value: ~
    kind: "}"
  - span:
      start: 213
      end: 216
    file_id: 1
    value: ~
    kind: fun
  - span:
      start: 217
      end: 220
    file_id: 1
    value:
      String: 9
    kind: identifier
  - span:
      start: 220
      end: 221
    file_id: 1
    value: ~
    kind: (
  - span:
      start: 221
      end: 225
    file_id: 1
    value: ~
    kind: self
  - span:
      start: 226
      end: 227
    file_id: 1
    value: ~
    kind: "@"
  - span:
      start: 227
      end: 231
    file_id: 1
    value:
      String: 6
    kind: identifier
  - span:
      start: 231
      end: 232
    file_id: 1
    value: ~
    kind: )
  - span:
      start: 233
      end: 234
    file_id: 1
    value: ~
    kind: "@"
  - span:
      start: 234
      end: 237
    file_id: 1
    value: ~
    kind: i64
  - span:
      start: 238
      end: 239
    file_id: 1
    value: ~
    kind: "{"
  - span:
      start: 241
      end: 245
    file_id: 1
    value: ~
    kind: self
  - span:
      start: 245
      end: 246
    file_id: 1
    value: ~
    kind: "."
  - span:
      start: 246
      end: 251
    file_id: 1
    value:
      String: 7
    kind: identifier
  - span:
      start: 252
      end: 253
    file_id: 1
    value: ~
    kind: +
  - span:
      start: 254
      end: 258
    file_id: 1
    value: ~
    kind: self
  - span:
      start: 258
      end: 259
    file_id: 1
    value: ~
    kind: "."
  - span:
      start: 259
      end: 264
    file_id: 1
    value:
      String: 8
    kind: identifier
  - span:
      start: 265
      end: 266
    file_id: 1
    value: ~
    kind: +
  - span:
      start: 267
      end: 271
    file_id: 1
    value: ~
    kind: self
  - span:
      start: 271
      end: 272
    file_id: 1
    value: ~
    kind: "."
  - span:
      start: 272
      end: 275
    file_id: 1
    value:
      String: 2
    kind: identifier
  - span:
      start: 275
      end: 276
    file_id: 1
    value: ~
    kind: "."
  - span:
      start: 276
      end: 279
    file_id: 1
    value:
      String: 2
    kind: identifier
  - span:
      start: 280
      end: 281
    file_id: 1
    value: ~
    kind: +
  - span:
      start: 282
      end: 286
    file_id: 1
    value: ~
    kind: self
  - span:
      start: 286
      end: 287
    file_id: 1
    value: ~
    kind: "."
  - span:
      start: 287
      end: 290
    file_id: 1
    value:
      String: 2
    kind: identifier
  - span:
      start: 290
      end: 291
    file_id: 1
    value: ~
    kind: "."
  - span:
      start: 291
      end: 294
    file_id: 1
    value:
      String: 3
    kind: identifier
  - span:
      start: 295
      end: 296
    file_id: 1
    value: ~
    kind: "}"
  - span:
      start: 298
      end: 301
    file_id: 1
    value: ~
    kind: fun
  - span:
      start: 302
      end: 306
    file_id: 1
    value:
      String: 10
    kind: identifier
  - span:
      start: 306
      end: 307
    file_id: 1
    value: ~
    kind: (
  - span:
      start: 307
      end: 308
    file_id: 1
    value: ~
    kind: )
  - span:
      start: 309
      end: 310
    file_id: 1
    value: ~
    kind: "@"
  - span:
      start: 310
      end: 313
    file_id: 1
    value: ~
    kind: i32
  - span:
      start: 314
      end: 315
    file_id: 1
    value: ~
    kind: "{"
  - span:
      start: 317
      end: 320
    file_id: 1
    value: ~
    kind: let
  - span:
      start: 321
      end: 325
    file_id: 1
    value:
      String: 11
    kind: identifier
  - span:
      start: 326
      end: 327
    file_id: 1
    value: ~
    kind: "@"
  - span:
      start: 327
      end: 330
    file_id: 1
    value: ~
    kind: i32
  - span:
      start: 331
      end: 332
    file_id: 1
    value: ~
    kind: "="
  - span:
      start: 333
      end: 334
    file_id: 1
    value:
      Integer: 0
    kind: int
  - span:
      start: 334
      end: 335
    file_id: 1
    value: ~
    kind: ;
  - span:
      start: 337
      end: 341
    file_id: 1
    value:
      String: 11
    kind: identifier
  - span:
      start: 342
      end: 343
    file_id: 1
    value: ~
    kind: "}"
interner:
  - Foo
  - foo
  - bar
  - init
  - void
  - Test
  - hello
  - world
  - add
  - main
  - test

